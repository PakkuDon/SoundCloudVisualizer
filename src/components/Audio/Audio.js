import React, { useEffect, useRef } from "react"
import PropTypes from "prop-types"

function Audio(props) {
  const audioElement = useRef(null)

  useEffect(() => {
    const audioContext = new (window.AudioContext ||
      window.webkitAudioContext)()
    if (audioContext) {
      const source = audioContext.createMediaElementSource(audioElement.current)
      const analyser = audioContext.createAnalyser()
      source.connect(analyser)
      analyser.connect(audioContext.destination)
      analyser.fftSize = 256
      props.onRender(analyser)
    }
  }, [])

  const { onEnded, src } = props

  return (
    <audio
      ref={audioElement}
      src={src}
      autoPlay
      crossOrigin="anonymous"
      controls
      onEnded={onEnded}
    >
      Your browser does not support the HTML5 audio element.
    </audio>
  )
}

Audio.propTypes = {
  src: PropTypes.string,
  onEnded: PropTypes.func,
  onRender: PropTypes.func,
}

Audio.defaultProps = {
  onEnded: () => {},
  onRender: () => {},
}

export default Audio
