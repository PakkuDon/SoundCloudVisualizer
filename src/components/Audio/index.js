import React from 'react'
import PropTypes from 'prop-types'

class Audio extends React.Component {
  constructor(props) {
    super(props)
    this.audioRef = React.createRef()
  }

  componentDidMount() {
    const audioContext = new (window.AudioContext || window.webkitAudioContext)
    if (audioContext) {
      const source = audioContext.createMediaElementSource(this.audioRef.current)
      const analyser = audioContext.createAnalyser()
      source.connect(analyser)
      analyser.connect(audioContext.destination)
      this.props.onRender(analyser)
    }
  }

  render() {
    const { src } = this.props

    return (
      <audio ref={this.audioRef} src={src} autoPlay crossOrigin="anonymous" controls>
        Your browser does not support the HTML5 audio element.
      </audio>
    )
  }
}

Audio.propTypes = {
  src: PropTypes.string,
  onRender: PropTypes.func,
}

Audio.defaultProps = {
  onRender: () => {},
}

export default Audio
