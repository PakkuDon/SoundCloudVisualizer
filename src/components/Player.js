import React from 'react'
import PropTypes from 'prop-types'
import { SoundCloudTrack } from '../propTypes'

const Player = ({ inputUrl, track }) => (
  <React.Fragment>
    <div>
      <form id="player-form">
        <input
          type="text"
          id="track-url"
          size="40"
          value={inputUrl}
        />
        <button type="button">Load song</button>
      </form>
    </div>
    <div id="song-information">
      <img id="thumbnail" src={track.artwork_url} />
      <div id="details">
        <div id="artist">{track && track.user.username}</div>
        <div id="title">{track.title}</div>
        <div id="genre">{track.genre}</div>
      </div>
    </div>
    <audio id="player" autoplay crossorigin="anonymous" controls src={track.stream_url} >
      Your browser does not support the HTML5 audio element.
    </audio>
    <footer>
      <a href="https://github.com/PakkuDon/SoundCloudVisualizer">Source code on GitHub</a>
    </footer>
  </React.Fragment>
)

Player.propTypes = {
  inputUrl: PropTypes.string,
  track: SoundCloudTrack,
}

export default Player
