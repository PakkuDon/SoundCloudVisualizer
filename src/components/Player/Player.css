.root {
  background-color: rgba(0, 160, 230, 0.1);
  width: 100%;
  text-align: center;
  padding: 1em;
}

.input {
  width: 30%;
}

.footer {
  text-align: right;
}
