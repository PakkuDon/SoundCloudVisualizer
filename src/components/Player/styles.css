.root {
  background-color: rgba(0, 160, 230, 0.1);
  width: 100%;
  text-align: center;
}

.thumbnail,
.details {
  display: inline-block;
}

.thumbnail {
  width: 100px;
  height: 100px;
}

.details {
  text-align: left;
}

.footer {
  text-align: right;
}
