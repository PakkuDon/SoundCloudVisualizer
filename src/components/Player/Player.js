import React, { useCallback } from "react"
import PropTypes from "prop-types"
import Button from "../Button"
import Audio from "../Audio"
import TrackInformation from "../TrackInformation"
import { SoundCloudTrack } from "../../propTypes"
import styles from "./Player.css"

const Player = ({
  errorMessage,
  inputUrl,
  onAudioEnded,
  onAudioRender,
  onSongSelect,
  onUrlEdit,
  track,
}) => {
  const onSubmit = useCallback(event => {
    event.preventDefault()
    onSongSelect()
  })

  return (
    <div className={styles.root}>
      <form onSubmit={onSubmit}>
        <label htmlFor="track-input">
          SoundCloud track URL
          <input
            id="track-input"
            className={styles.input}
            type="text"
            value={inputUrl}
            onChange={event => onUrlEdit(event.target.value)}
          />
        </label>
        <Button type="submit">Load song</Button>
      </form>
      <div>{errorMessage}</div>
      {track && <TrackInformation track={track} />}
      <Audio
        src={track && track.stream_url}
        onEnded={onAudioEnded}
        onRender={onAudioRender}
      />
      <footer className={styles.footer}>
        <a href="https://github.com/PakkuDon/SoundCloudVisualizer">
          Source code on GitHub
        </a>
      </footer>
    </div>
  )
}

Player.propTypes = {
  errorMessage: PropTypes.string,
  inputUrl: PropTypes.string,
  track: SoundCloudTrack,
  onAudioRender: PropTypes.func,
  onSongSelect: PropTypes.func,
  onUrlEdit: PropTypes.func,
}

Player.defaultProps = {
  errorMessage: "",
  inputUrl: "",
  onAudioRender: () => {},
  onUrlEdit: () => {},
}

export default Player
