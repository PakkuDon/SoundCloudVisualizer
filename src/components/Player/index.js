import React from 'react'
import PropTypes from 'prop-types'
import Button from '../Button'
import { SoundCloudTrack } from '../../propTypes'
import styles from './styles.css'

const Player = ({ inputUrl, track }) => (
  <React.Fragment>
    <div>
      <form>
        <input
          type="text"
          size="40"
          value={inputUrl}
        />
      <Button>Load song</Button>
      </form>
    </div>
    <div>
      <img className={styles.thumbnail} src={track.artwork_url} />
      <div className={styles.details}>
        <div>{track && track.user.username}</div>
        <div>{track.title}</div>
        <div>{track.genre}</div>
      </div>
    </div>
    <audio autoplay crossorigin="anonymous" controls src={track.stream_url} >
      Your browser does not support the HTML5 audio element.
    </audio>
    <div className={styles.footer}>
      <a href="https://github.com/PakkuDon/SoundCloudVisualizer">Source code on GitHub</a>
    </div>
  </React.Fragment>
)

Player.propTypes = {
  inputUrl: PropTypes.string,
  track: SoundCloudTrack,
}

export default Player
