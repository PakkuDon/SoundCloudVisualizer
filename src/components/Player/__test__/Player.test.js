import React from "react"
import { shallow } from "enzyme"

import Player from "../Player"
import Audio from "../../Audio"

const createTrack = () => ({
  id: 1,
  title: "Hello World",
  genre: "",
  permalink_url: "",
  stream_url: "",
  artwork_url: "",
  user: {
    username: "Bob",
  },
})

describe("Player", () => {
  it("renders track details", () => {
    const track = createTrack()
    const wrapper = shallow(<Player track={track} />)

    expect(wrapper).toMatchSnapshot()
  })

  describe("on url edit", () => {
    it("calls supplied onUrlEdit handler", () => {
      const spy = jest.fn()
      const track = createTrack()
      const wrapper = shallow(<Player track={track} onUrlEdit={spy} />)

      wrapper.find("input").simulate("change", { target: {} })
      expect(spy).toBeCalled()
    })
  })

  describe("on form submission", () => {
    it("calls supplied onSongSelect handler", () => {
      const spy = jest.fn()
      const track = createTrack()
      const wrapper = shallow(<Player track={track} onSongSelect={spy} />)

      wrapper.find("form").simulate("submit", { preventDefault: () => {} })
      expect(spy).toBeCalled()
    })
  })

  describe("on audio ended", () => {
    it("calls supplied onAudioEnded handler", () => {
      const spy = jest.fn()
      const track = createTrack()
      const wrapper = shallow(<Player track={track} onAudioEnded={spy} />)

      wrapper.find(Audio).prop("onEnded")()
      expect(spy).toBeCalled()
    })
  })

  describe("on audio render", () => {
    it("calls supplied onAudioRender handler", () => {
      const spy = jest.fn()
      const track = createTrack()
      const wrapper = shallow(<Player track={track} onAudioRender={spy} />)

      wrapper.find(Audio).prop("onRender")()
      expect(spy).toBeCalled()
    })
  })
})
