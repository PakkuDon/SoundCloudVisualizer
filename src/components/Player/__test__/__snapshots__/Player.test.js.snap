// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Player renders track details 1`] = `
<div>
  <div
    class="root"
  >
    <form>
      <label
        for="track-input"
      >
        SoundCloud track URL
        <input
          class="input"
          id="track-input"
          type="text"
          value=""
        />
      </label>
      <button
        class="root"
        title=""
        type="submit"
      >
        Load song
      </button>
    </form>
    <div />
    <div
      class="root"
    >
      <img
        alt="Hello World artwork"
        class="thumbnail"
        src=""
      />
      <div
        class="details"
      >
        <div>
          Bob
        </div>
        <div>
          Hello World
        </div>
        <div />
      </div>
    </div>
    <audio
      autoplay=""
      controls=""
      crossorigin="anonymous"
      src=""
    >
      Your browser does not support the HTML5 audio element.
    </audio>
    <footer
      class="footer"
    >
      <a
        href="https://github.com/PakkuDon/SoundCloudVisualizer"
      >
        Source code on GitHub
      </a>
    </footer>
  </div>
</div>
`;
